onnxruntime/core/optimizer/constant_sharing.cc

```diff
@@ -0,0 +1,233 @@
+    if (std::find(consumers.begin(), consumers.end(), node) == consumers.end()) {
+      graph.AddConsumerNode(shared_initializer_node_arg->Name(), node);
+    }
+  }
+
+  // Remove the initializer if no other consumer nodes.
+  if (graph.GetConsumerNodes(origin_initializer_node_arg->Name()).size() == 0) {
+    graph.RemoveInitializedTensor(origin_initializer_node_arg->Name());
+  }
+}
+
+int32_t GenerateUniqueValueId() {
+  static int32_t value_key_generator = 1;
+  return value_key_generator++;
+}
+
+}  // namespace
+
+Status ConstantSharing::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/,
+                                  const logging::Logger& /*logger*/) const {
+  // Accumulated map from type/value/rank to initializer:
+  //   The key is a string representation of initializer's data type, value and rank.
+  //   The value is newly created initializer NodeArg* to be shared.
+  std::map<std::string, NodeArg*> pattern_key_to_shared_arg_map;
+  const InitializedTensorSet& initialized_tensor_set = graph.GetAllInitializedTensors();
+  InlinedVector<std::string> original_initializer_names;
+  original_initializer_names.reserve(initialized_tensor_set.size());
+  for (const auto& entry : initialized_tensor_set) {
+    // Ignore if the initializer already handled, or not a constant initializer.
+    if (IsSharedInitializer(entry.first) || !graph_utils::IsConstantInitializer(graph, entry.first)) {
+      continue;
+    }
+
+    if (excluded_initializers_.find(entry.first) != excluded_initializers_.end()) {
+      continue;
+    }
+
+    original_initializer_names.push_back(entry.first);
+  }
+
+  // We avoid using the scalar value directly in pattern_key because the value for example INT_MAX can be super big
+  // and it will be hard to read. Instead, we use a unique id for each scalar value, and a map from the value to
+  // unique id.
+  InlinedHashMap<int32_t, int32_t> int32_value_to_value_id_map;
+  InlinedHashMap<int64_t, int32_t> int64_value_to_value_id_map;
+  InlinedHashMap<float, int32_t> float_value_to_value_id_map;
+  InlinedHashMap<float, int32_t> half_value_to_value_id_map;
+  InlinedHashMap<double, int32_t> double_value_to_value_id_map;
+  for (const auto& initializer_name : original_initializer_names) {
+    NodeArg* origin_initializer_node_arg = graph.GetNodeArg(initializer_name);
+
+    // Already handled, skip it.
+    if (origin_initializer_node_arg == nullptr) {
+      continue;
+    }
+
+    auto input_shape = origin_initializer_node_arg->Shape();
+    if (input_shape == nullptr || !IsSingleValueShape(input_shape)) {
+      continue;
+    }
+
+    // Ignore if not constant initializers.
+    const ONNX_NAMESPACE::TensorProto* tensor_proto = graph.GetConstantInitializer(origin_initializer_node_arg->Name(),
+                                                                                   true);
+
+    int32_t data_type = tensor_proto->data_type();
```

Comment:
we can be sure here that the tensor_proto is not a nullptr, right? (it is a runtime constant initializer?).

Reply:
Yeah, there is a constant initializer check when we get original_initializer_names in  the beginning of this function, while it is still not check the null directly. I added the null check here.